enterprise/backend/src/metabase_enterprise/serialization/api.clj

```diff
@@ -0,0 +1,242 @@
+  (swap! callbacks* update req conj cb))
+
+(defn- run-callbacks! [req]
+  (run! #(%) (get @callbacks* req))
+  (swap! callbacks* dissoc req))
+
+(defn- ba-copy [f]
+  (let [baos (ByteArrayOutputStream.)]
+    (io/copy f baos)
+    (.toByteArray baos)))
+
+(defn- on-response! [req data]
+  (reify
+    ;; Real HTTP requests and mt/user-real-request go here
+    ring.protocols/StreamableResponseBody
+    (write-body-to-stream [_ response out]
+      (ring.protocols/write-body-to-stream data response out)
+      (run-callbacks! req))
+
+    ;; mt/user-http-request goes here
+    clojure.java.io.IOFactory
+    (make-input-stream [_ _]
+      (let [res (io/input-stream (if (instance? File data)
+                                   (ba-copy data)
+                                   data))]
+        (run-callbacks! req)
+        res))))
+
+;;; Storage
+
+(def parent-dir "Dir for storing serialization API export-in-progress and archives."
+  (let [f (io/file (System/getProperty "java.io.tmpdir") "serdesv2-api")]
+    (.mkdirs f)
+    (.deleteOnExit f)
+    (.getPath f)))
+
+(defn- compress-tgz [^File src ^File dst]
+  (when-not (.exists src)
+    (throw (ex-info (format "Path is not readable or does not exist: %s" src)
+                    {:path src})))
+  (let [prefix (.getPath (.getParentFile src))]
+    (with-open [tar (-> (io/output-stream dst)
+                        (GzipCompressorOutputStream. (doto (GzipParameters.)
+                                                       (.setModificationTime (System/currentTimeMillis))))
+                        (TarArchiveOutputStream. 512 "UTF-8"))]
+      (.setLongFileMode tar TarArchiveOutputStream/LONGFILE_POSIX)
+
+      (doseq [^File f (file-seq src)
+              :let    [path-in-tar (subs (.getPath f) (count prefix))
+                       entry (TarArchiveEntry. f path-in-tar)]]
+        (.putArchiveEntry tar entry)
+        (when (.isFile f)
+          (with-open [s (io/input-stream f)]
+            (io/copy s tar)))
+        (.closeArchiveEntry tar)))
+    dst))
+
+(defn- entries [^TarArchiveInputStream tar]
+  (lazy-seq
+   (when-let [entry (.getNextEntry tar)]
+     (cons entry (entries tar)))))
+
+(defn- uncompress-tgz [^File archive ^File dst]
+  (let [dir-name (atom nil)]
+    (with-open [tar (-> (io/input-stream archive)
+                        (GzipCompressorInputStream.)
+                        (TarArchiveInputStream.))]
+      (doseq [^TarArchiveEntry e (entries tar)]
+        (when-not @dir-name
+          (reset! dir-name (.getName e)))
+        (let [f (io/file dst (.getName e))]
+          (if (.isFile e)
+            (io/copy tar f)
+            (.mkdirs f)))))
+    @dir-name))
+
+;;; Logging
+
+(defn- find-logger-layout
+  "Find any logger with a specified layout"
+  [^LoggerConfig logger]
+  (when logger
+    (or (first (keep #(.getLayout ^AbstractAppender (val %)) (.getAppenders logger)))
+        (recur (.getParent logger)))))
+
+(defn- make-logger ^AutoCloseable [ns ^File f]
```

Comment:
I've been thinking about moving this to `metabase.logger` (`metabase.util.log` is both clj/cljs, so it's not applicable there), but maybe it's better done when it's needed somewhere else.

OTOH maybe this can be re-used in tests...